File: flutter/ios/Classes/DigiaSuppressPlugin.m
//
//  DigiaSuppressPlugin.m
//  digia_webengage_plugin
//

#import "DigiaSuppressPlugin.h"
#import <objc/message.h>

// Valid command values that identify a Digia nudge campaign.
static NSString * const kCommandShowDialog       = @"SHOW_DIALOG";
static NSString * const kCommandShowBottomSheet  = @"SHOW_BOTTOM_SHEET";

// MethodChannel name — must match Dart and Android.
static NSString * const kChannelName = @"plugins.digia.tech/webengage_suppress";

@interface DigiaSuppressPlugin ()
@property (nonatomic, strong, nullable) FlutterMethodChannel *channel;
@end

@implementation DigiaSuppressPlugin

// MARK: - Shared Instance

static DigiaSuppressPlugin *_shared = nil;

+ (nullable DigiaSuppressPlugin *)shared {
    return _shared;
}

// MARK: - FlutterPlugin Registration

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    DigiaSuppressPlugin *instance = [[DigiaSuppressPlugin alloc] init];
    _shared = instance;

    FlutterMethodChannel *channel =
        [FlutterMethodChannel methodChannelWithName:kChannelName
                                    binaryMessenger:[registrar messenger]];
    [registrar addMethodCallDelegate:instance channel:channel];
    instance.channel = channel;
    // NOTE: Do NOT set notificationDelegate here via KVC.
    // Pass +shared via WebEngageConfig in AppDelegate after registering plugins.
}

- (void)detachFromEngineForRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    self.channel = nil;
    if (_shared == self) {
        _shared = nil;
    }
}

// MARK: - FlutterMethodCallDelegate

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([call.method isEqualToString:@"trackSystemEvent"]) {
        NSDictionary *args       = call.arguments;
        NSString     *eventName  = args[@"eventName"];
        NSDictionary *systemData = args[@"systemData"] ?: @{};
        NSDictionary *eventData  = args[@"eventData"]  ?: @{};
        if (eventName.length) {
            // Build the andValue dict the same way WebEngage fires notification_view
            // internally: put campaign data under "system_data_overrides" so the
            // WEGEvent factory merges it into system_data (not event_data).
            NSMutableDictionary *andValue = [NSMutableDictionary dictionary];
            if (systemData.count) {
                andValue[@"system_data_overrides"] = systemData;
            }
            if (eventData.count) {
                andValue[@"event_data_overrides"] = eventData;
            }

            NSLog(@"[DigiaSuppressPlugin] trackSystemEvent — name=%@ andValue=%@", eventName, andValue);

            // Mark sessionCloses BEFORE firing the event so re-evaluation
            // triggered by the event already sees the mark.
            if ([eventName isEqualToString:@"notification_close"]) {
                NSString *expId = systemData[@"experiment_id"];
                if (expId.length > 0) {
                    Class rendererCls = NSClassFromString(@"WEGRenderer");
                    SEL sharedSel2 = NSSelectorFromString(@"sharedInstance");
                    SEL closesSel  = NSSelectorFromString(@"sessionCloses");
                    if (rendererCls && [rendererCls respondsToSelector:sharedSel2]) {
                        id renderer = ((id (*)(Class, SEL))objc_msgSend)(rendererCls, sharedSel2);
                        if (renderer && [renderer respondsToSelector:closesSel]) {
                            NSMutableDictionary *closes = ((id (*)(id, SEL))objc_msgSend)(renderer, closesSel);
                            if ([closes isKindOfClass:[NSMutableDictionary class]]) {
                                NSString *closeKey = [expId stringByAppendingString:@"_close"];
                                closes[closeKey] = @1;
                                NSLog(@"[DigiaSuppressPlugin] Marked %@ in sessionCloses: %@", closeKey, closes);
                            }
                        }
                    }
                }
            }

            // Use trackSDKEventWithName:andValue: which bypasses reserved-name
            // guards and routes through the full SDK event pipeline.
            SEL sharedSel = NSSelectorFromString(@"sharedInstance");
            SEL trackSel  = NSSelectorFromString(@"trackSDKEventWithName:andValue:");

            Class analyticsCls = NSClassFromString(@"WEGAnalyticsImpl");
            if (analyticsCls) {
                typedef id  (*SharedFn)(Class, SEL);
                typedef void (*TrackFn)(id, SEL, NSString *, NSDictionary *);

                id analytics = ((SharedFn)objc_msgSend)(analyticsCls, sharedSel);
                if (analytics) {
                    ((TrackFn)objc_msgSend)(analytics, trackSel, eventName, andValue);
                    NSLog(@"[DigiaSuppressPlugin] trackSystemEvent — fired: %@", eventName);
                } else {
                    NSLog(@"[DigiaSuppressPlugin] trackSystemEvent — analytics not available");
                }
            }

        } else {
            NSLog(@"[DigiaSuppressPlugin] trackSystemEvent: skipped — eventName is empty");
        }
        result(nil);
    } else {
        // `configure` was previously used to toggle suppressRendering globally.
        // Suppression is now handled per-campaign in notificationPrepared:shouldStop:.
        result(nil);
    }
}

// MARK: - WEGInAppNotificationProtocol

- (NSDictionary *)notificationPrepared:(NSDictionary<NSString *, id> *)inAppNotificationData
                             shouldStop:(BOOL *)stopRendering {

    // ── Step 1: Identify campaign type ──────────────────────────────────────
    if (![self isDigiaCampaign:inAppNotificationData]) {
        NSLog(@"[DigiaSuppressPlugin] notificationPrepared — non-Digia campaign, allowing normal rendering");
        // Normal campaign → allow WebEngage SDK to render it as usual.
        return inAppNotificationData;
    }
    // ── Step 2: Digia campaign → suppress WebEngage's own renderer ──────────
    *stopRendering = YES;

    // ── Step 3: Ensure experimentId is forwarded under the canonical key ─────
    NSString *experimentId = inAppNotificationData[@"notificationEncId"]
                          ?: inAppNotificationData[@"experimentId"]
                          ?: inAppNotificationData[@"experiment_id"];
    NSMutableDictionary *payload = [inAppNotificationData mutableCopy];
    if (experimentId && !payload[@"experimentId"]) {
        payload[@"experimentId"] = experimentId;
    }

    // ── Step 5: Send data to Flutter on the main thread ──────────────────────
    FlutterMethodChannel *channel = self.channel;
    NSDictionary *immutablePayload = [payload copy];
    dispatch_async(dispatch_get_main_queue(), ^{
        [channel invokeMethod:@"onInAppPrepared" arguments:immutablePayload];
    });

    return inAppNotificationData;
}

- (void)notificationDismissed:(NSDictionary<NSString *, id> *)inAppNotificationData {
    NSString *experimentId = inAppNotificationData[@"notificationEncId"]
                          ?: inAppNotificationData[@"experimentId"]
                          ?: inAppNotificationData[@"experiment_id"];

    FlutterMethodChannel *channel = self.channel;
    NSDictionary *args = @{ @"experimentId": experimentId ?: [NSNull null] };
    dispatch_async(dispatch_get_main_queue(), ^{
        [channel invokeMethod:@"onInAppDismissed" arguments:args];
    });
}

// MARK: - Campaign Type Identification

/// Returns YES when the notification data carries a Digia rendering contract.
///
/// Mirrors the Dart WebEngagePayloadMapper._normalizeWithDigiaContract logic:
///   1. Structured keys: viewId present + at least one routing signal (type or known command).
///   2. Embedded HTML: any string value contains <digia or digia-payload.
- (BOOL)isDigiaCampaign:(NSDictionary<NSString *, id> *)data {
    return [self hasDigiaContractKeysInData:data] || [self containsDigiaHtmlInNode:data];
}

/// Checks for the structured contract: viewId + routing signal at the top level.
- (BOOL)hasDigiaContractKeysInData:(NSDictionary<NSString *, id> *)data {
    NSString *viewId = [data[@"viewId"] isKindOfClass:[NSString class]]
                     ? [data[@"viewId"] stringByTrimmingCharactersInSet:NSCharacterSet.whitespaceCharacterSet]
                     : nil;
    if (!viewId.length) return NO;

    NSString *type    = [data[@"type"]    isKindOfClass:[NSString class]] ? data[@"type"]    : nil;
    NSString *command = [data[@"command"] isKindOfClass:[NSString class]] ? data[@"command"] : nil;

    if (type) return YES;
    NSString *upperCommand = [command uppercaseString];
    if ([upperCommand isEqualToString:kCommandShowDialog] ||
        [upperCommand isEqualToString:kCommandShowBottomSheet]) return YES;

    return NO;
}

/// Recursively scans all string values for embedded Digia HTML markers.
- (BOOL)containsDigiaHtmlInNode:(id)node {
    if ([node isKindOfClass:[NSString class]]) {
        NSString *lower = [(NSString *)node lowercaseString];
        return [lower containsString:@"<digia"] || [lower containsString:@"digia-payload"];
    }
    if ([node isKindOfClass:[NSDictionary class]]) {
        for (id value in [(NSDictionary *)node allValues]) {
            if ([self containsDigiaHtmlInNode:value]) return YES;
        }
    }
    if ([node isKindOfClass:[NSArray class]]) {
        for (id item in (NSArray *)node) {
            if ([self containsDigiaHtmlInNode:item]) return YES;
        }
    }
    return NO;
}

@end

File: flutter/ios/Classes/DigiaSuppressPlugin.swift
import Flutter
import WebEngage

/// Flutter plugin that acts as the WebEngage `WEGInAppNotificationProtocol` delegate on iOS.
///
/// **Campaign routing flow:**
/// ```
/// Native Layer (iOS)
///     ↓
/// notificationPrepared() called
///     ↓
/// isDigiaCampaign(data)?
///     ↓ YES                         ↓ NO
/// suppressRendering = true      suppressRendering stays false
/// deduplicate by experimentId   return data as-is
/// send data → Flutter            (WebEngage SDK renders normally)
/// return data
/// ```
///
/// A campaign is identified as "Digia" when its notification data either:
///   - contains a `viewId` key plus a routing signal (`type` or a valid `command`), OR
///   - embeds a `<digia .../>` or `<script digia-payload>` HTML tag anywhere in its values.
///
/// The plugin instance is stored as a static `shared` reference populated during
/// `register(with:)`. Pass `DigiaSuppressPlugin.shared` as the `notificationDelegate`
/// inside `WebEngageConfig` **after** calling `GeneratedPluginRegistrant.register(with:)`
/// so the instance is available. See `AppDelegate.swift`.
public final class DigiaSuppressPlugin: NSObject, FlutterPlugin, WEGInAppNotificationProtocol {

    // MARK: - Shared Instance

    /// The plugin instance created during Flutter plugin registration.
    /// Available after `GeneratedPluginRegistrant.register(with:)` is called.
    @objc public private(set) static var shared: DigiaSuppressPlugin?

    // MARK: - Private State

    private var channel: FlutterMethodChannel?

    /// Tracks the experimentId of the last Digia in-app forwarded to Dart.
    /// WebEngage evaluates campaigns on every screen navigation, producing repeated
    /// `notificationPrepared` callbacks. Only the first is forwarded; cleared on dismiss.
    private var activeExperimentId: String?

    // MARK: - FlutterPlugin Registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = DigiaSuppressPlugin()
        DigiaSuppressPlugin.shared = instance

        let channel = FlutterMethodChannel(
            name: "plugins.digia.tech/webengage_suppress",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: channel)
        instance.channel = channel
        // NOTE: Do NOT attempt to set the WebEngage notificationDelegate here via KVC.
        // The WebEngage SDK has no public setter for notificationDelegate post-init.
        // Pass DigiaSuppressPlugin.shared via WebEngageConfig in AppDelegate instead.
    }

    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        channel = nil
        activeExperimentId = nil
        if DigiaSuppressPlugin.shared === self {
            DigiaSuppressPlugin.shared = nil
        }
    }

    // MARK: - FlutterMethodCallDelegate

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        // `configure` was previously used to toggle suppressRendering globally.
        // Suppression is now handled per-campaign in notificationPrepared(_:shouldStop:),
        // so this call is accepted but has no effect.
        result(nil)
    }

    // MARK: - WEGInAppNotificationProtocol

    public func notificationPrepared(
        _ inAppNotificationData: [String: Any],
        shouldStop stopRendering: UnsafeMutablePointer<ObjCBool>
    ) -> [String: Any] {

        // ── Step 1: Identify campaign type ──────────────────────────────────
        guard isDigiaCampaign(inAppNotificationData) else {
            // Normal campaign → allow WebEngage SDK to render it as usual.
            return inAppNotificationData
        }

        // ── Step 2: Digia campaign → suppress WebEngage's own renderer ──────
        stopRendering.pointee = true

        // ── Step 3: Deduplicate repeated callbacks for the same campaign ─────
        let experimentId = inAppNotificationData["experimentId"] as? String
                        ?? inAppNotificationData["experiment_id"] as? String

        if let eid = experimentId, eid == activeExperimentId {
            return inAppNotificationData
        }
        activeExperimentId = experimentId

        // ── Step 4: Ensure experimentId is always forwarded under canonical key
        var payload = inAppNotificationData
        if let eid = experimentId, payload["experimentId"] == nil {
            payload["experimentId"] = eid
        }

        // ── Step 5: Send data to Flutter ────────────────────────────────────
        channel?.invokeMethod("onInAppPrepared", arguments: payload)
        return inAppNotificationData
    }

    public func notificationDismissed(_ inAppNotificationData: [String: Any]) {
        let experimentId = inAppNotificationData["experimentId"] as? String
                        ?? inAppNotificationData["experiment_id"] as? String

        if let eid = experimentId, eid == activeExperimentId {
            activeExperimentId = nil
        }

        channel?.invokeMethod("onInAppDismissed", arguments: ["experimentId": experimentId as Any])
    }

    // MARK: - Campaign Type Identification

    /// Returns `true` when the notification data carries a Digia rendering contract.
    ///
    /// Mirrors the Dart `WebEngagePayloadMapper._normalizeWithDigiaContract` logic:
    /// 1. **Top-level keys**: `viewId` is present **and** at least one routing signal
    ///    (`type` or a known `command`) exists.
    /// 2. **Embedded HTML**: any string value in the dictionary contains a
    ///    `<digia .../>` attribute tag or a `<script digia-payload>` block.
    private func isDigiaCampaign(_ data: [String: Any]) -> Bool {
        if hasDigiaContractKeys(in: data) { return true }
        if containsDigiaHtml(in: data)    { return true }
        return false
    }

    /// Checks for the structured contract: `viewId` + routing signal at the top level.
    private func hasDigiaContractKeys(in data: [String: Any]) -> Bool {
        let viewId = (data["viewId"] as? String)?.trimmingCharacters(in: .whitespaces) ?? ""
        guard !viewId.isEmpty else { return false }

        let type    = (data["type"] as? String)?.lowercased()
        let command = (data["command"] as? String)?.uppercased()

        if type != nil { return true }
        if command == "SHOW_DIALOG" || command == "SHOW_BOTTOM_SHEET" { return true }
        return false
    }

    /// Recursively scans all string values for embedded Digia HTML markers.
    private func containsDigiaHtml(in node: Any) -> Bool {
        if let str = node as? String {
            let lower = str.lowercased()
            return lower.contains("<digia") || lower.contains("digia-payload")
        }
        if let dict = node as? [String: Any] {
            return dict.values.contains(where: containsDigiaHtml)
        }
        if let array = node as? [Any] {
            return array.contains(where: containsDigiaHtml)
        }
        return false
    }
}
